src/pudl/transform/ferc1.py

```diff
@@ -694,6 +694,106 @@ def combine_axis_columns_xbrl(
     return df
 
 
+class CheckTableCalculations(TransformParams):
+    """Parameters for checking xbrl-metadata based calculations within a table."""
+
+    column_to_check: str | None = None
+    """Name of data column to check.
+
+    This will typically be ``dollar_amount`` or ``ending_balance`` in hte
+    """
+    calculation_tolerance: float = 0.05
+    """The tolerance ratio of the off calcuations and the possible calcuations."""
+
+
+def check_table_calculations(
+    df: pd.DataFrame,
+    tbl_meta: pd.DataFrame,
+    xbrl_factoid_name: str,
+    table_name: str,
+    params: CheckTableCalculations,
+) -> pd.DataFrame:
+    """Calculate the intra-table calculations and ensure the table is within tolerance.
+
+    Args:
+        df: processed table.
+        tbl_meta: processed table xbrl metadata.
+        xbrl_factoid_name: column name of the XBRL factoid in the processed table.
+        table_name: name of the PUDL table.
+        params: :class:`CheckTableCalculations` parameters.
+    """
+    # skip the calculations that have any components from other tables.
+    inter_table_calculated_values = list(
+        tbl_meta[
+            tbl_meta.calc_contains_components_from_other_tables
+        ].xbrl_factoid.unique()
+    )
+    if inter_table_calculated_values:
+        logger.info(
+            "Skipping calcuated values because they are inter-table calucations: "
+            f"{inter_table_calculated_values}"
+        )
+    intra_tbl_calcs = tbl_meta[
+        ~tbl_meta.calc_contains_components_from_other_tables
+        & (tbl_meta.row_type_xbrl == "calculated_value")
+    ]
+    pks = pudl.metadata.classes.Resource.from_id(table_name).schema.primary_key
+    pks_wo_factoid = [col for col in pks if col != xbrl_factoid_name]
+    calculated_dfs = []
+    for calcuated_factoid, calculation in zip(
+        intra_tbl_calcs.xbrl_factoid, intra_tbl_calcs.calculations
+    ):
+        calc_df = (
+            pd.merge(
+                df,
+                pd.DataFrame(literal_eval(calculation)),
```

Comment:
Why are we using `ast.literal_eval()` (which can evaluate any string as if it were python code) rather than the built-in JSON serialization / deserialization functions `json.load()` / `json.dump()`?  It seems like all of the interaction with the calculations and other metadata is about working with JSON rather than lower level Python code parsing.

Reply:
bc of the " / ' problem

Reply:
which you've mentioned before is an upstream problem that we should probably deal with. it'd be great to use the `json` tools directly! and this should be very straightforwad to swap out when there isn't mixed quotes.

Reply:
Okay that's what I thought. We should never be writing invalid JSON into these fields in the first place.  Do we know where that's happening? The JSON we're getting from the original extractor seems to be fine (it parses as good JSON in `json_normalize()`) so I would guess this is happening in `process_xbrl_metadata()` somewhere, Probably because we assign some dataframe cells to contain a python object rather than a JSON string representation of that object which would be generated by e.g. `json.dumps(my_object)` (the `s` is for string).

Reply:
A [JSON & Python tutorial](https://realpython.com/python-json/#python-supports-json-natively) from RealPython.

Reply:
converted!